lambdaclass/mirra_backend
# [GH-635] Configurator web app basis

## Disclaimer

The configurator app is not integrated with the game yet, that going to be addressed in the latest iterations since it implies a lot of changes and play testing.

Addresses the item "Display the config as key/value pairs. This can be in a tabular format similar to Excel or in another way." from #635 

## Motivation

Restructure the way we visualize and manage JSONs. To create an abstract configurator that doesn't rely on a specific game, we need to add some layers to achieve this. I will attach an image regarding this change.

<p align="center">
  <img src="https://github.com/lambdaclass/mirra_backend/assets/67444519/0c42af91-4cd6-4dc0-b549-56201bc8ce9e">
</div>

There a few concepts which I will explain below:
- Game: A game is an abstraction for every game we have, it is mainly used to group things which are called `Configuration Groups`
- Configuration Group: These are mainly used to group Configurations have a version control within our Configurations
- Configuration: These are mainly used to store the json data. 

For example, we could create something like this

A `game` named Champions of Mirra -> A `configuration group` named `Items` -> an have a lot of configurations

Here's a short video showing it

https://github.com/lambdaclass/mirra_backend/assets/67444519/4d837aa5-3930-4ee6-8366-76b79204c3f2

## Summary of changes
- Updated Phoenix LiveView version.
- Added an MVC for the new Game schema.
- Added an MVC for the new ConfigurationGroup schema
- Refactored the routing for the configurator app. Removed some endpoints that will be addressed in another iteration (Like setting the current configuration for a configuration group)
- Removed all styles and stuff related to Phoenix base setup
- Added and modified some fields from the Configuration Schema. 
    - Renamed `is_default` to `current`
    - Added a relationship between groups and configuration groups.
    - Added a name field
- Fixed the tests (We should address this)
- Added a condition that says if a map has less than two keys, we will display it as a plain text json.

## How to test it?

- Go to `localhost:4100/games` (We're going to add a LogIn in the future, that's the reason of the main endpoint)
- Create a game (name it as you want)
- Create a ConfigurationGroup (again, name it as you want)
- Create a configuration an add one of these jsons
You can try out some of these jsons ( (At the moment we should follow format of the ones below). In the next PR we may handle multiple main keys

<details>
<summary>Map with common keys</summary>

``` json
{
 "main_key1": {
    "key1": "value1",
    "key2": "value2",
    "key3": "value3",
    "key4": "value4",
    "key5": "value5"
 }
}
```
</details>

<details>
<summary>Map with lists</summary>

``` json
{
 "main_key1": {
    "key1": ["value1", "value2", "value3"],
    "key2": ["value4", "value5", "value6"],
    "key3": ["value7", "value8", "value9"],
    "key4": "value10",
    "key5": "value11"
 }
}
```
</details>

<details>
<summary>Nested maps</summary>

``` json
{
 "main_key1": {
    "sub_key1": {
        "sub_sub_key1": "value1",
        "sub_sub_key2": "value2"
    }
 }
}
```
</details>

You can also use configurations from `config.json` too, like this

<details>
<summary>Items</summary>

``` json
{
  "items": [
    {
      "name": "golden_clock",
      "effects": [
        {
          "stamina_faster": {
            "interval_decrease_by": 0.3,
            "duration_ms": 7000
          }
        }
      ]
    },
    {
      "name": "magic_boots",
      "effects": [
        {
          "speed_boost": {
            "amount": 9,
            "duration_ms": 10000
          }
        }
      ]
    },
    {
      "name": "mirra_blessing",
      "effects": [
        {
          "damage_immunity": {
            "duration_ms": 1
          }
        }
      ]
    }
  ]
 }
```
</details>

## Checklist
